Assets/Mirror/Runtime/NetworkReader.cs

```diff
@@ -10,51 +18,160 @@ namespace Mirror
-        public bool ReadBoolean() => reader.ReadBoolean();
-        public short ReadInt16() => reader.ReadInt16();
-        public ushort ReadUInt16() => reader.ReadUInt16();
-        public int ReadInt32() => reader.ReadInt32();
-        public uint ReadUInt32() => reader.ReadUInt32();
-        public long ReadInt64() => reader.ReadInt64();
-        public ulong ReadUInt64() => reader.ReadUInt64();
-        public decimal ReadDecimal() => reader.ReadDecimal();
-        public float ReadSingle() => reader.ReadSingle();
-        public double ReadDouble() => reader.ReadDouble();
+            other = ((ulong)ReadByte()) << 8;
+            value |= other;
+
+            other = ((ulong)ReadByte()) << 16;
+            value |= other;
+
+            other = ((ulong)ReadByte()) << 24;
+            value |= other;
+
+            other = ((ulong)ReadByte()) << 32;
+            value |= other;
+
+            other = ((ulong)ReadByte()) << 40;
+            value |= other;
+
+            other = ((ulong)ReadByte()) << 48;
+            value |= other;
+
+            other = ((ulong)ReadByte()) << 56;
+            value |= other;
+            return value;
+        }
+        public decimal ReadDecimal()
+        {
+            int[] bits = new int[4];
+
+            bits[0] = ReadInt32();
+            bits[1] = ReadInt32();
+            bits[2] = ReadInt32();
+            bits[3] = ReadInt32();
+
+            return new decimal(bits);
+        }
+        public float ReadSingle()
+        {
+            uint value = ReadUInt32();
+            return FloatConversion.ToSingle(value);
+        }
+        public double ReadDouble()
+        {
+            ulong value = ReadUInt64();
+            return FloatConversion.ToDouble(value);
+        }
 
         // note: this will throw an ArgumentException if an invalid utf8 string is sent
         // null support, see NetworkWriter
-        public string ReadString() => ReadBoolean() ? reader.ReadString() : null;
+        public string ReadString()
+        {
+            // isNull?
+            if (ReadBoolean())
+            {
+                // read number of bytes
+                UInt16 numBytes = ReadUInt16();
+                if (numBytes == 0)
+                    return "";
+
+                // make sure it's within limits to avoid allocation attacks etc.
+                if (numBytes >= stringBuffer.Length)
+                {
+                    throw new EndOfStreamException("ReadString() too long: " + numBytes + ". Limit is: " + stringBuffer.Length);
+                }
+
+                // read the bytes
+                ReadBytes(stringBuffer, numBytes);
+
+                // convert to string via encoding
+                return encoding.GetString(stringBuffer, 0, numBytes);
+            }
+            return null;
+        }
+
+        // read bytes into the passed buffer
+        public byte[] ReadBytes(byte[] bytes, int count)
+        {
+            if (Position + count > buffer.Count)
```

Comment:
this is fine,  but we could also get a non allocating readbytes like this:

```cs
public ArraySegment<byte> ReadSlice(int count) {...}
```
that can return an arraysegment of the buffer.    This would be useful for reading message payloads or rpc parameters without extra allocations

Reply:
great idea. probably best in a separate PR though

Reply:
yes,  sure  separate PR would be fine